text/0000-futures.md

```diff
@@ -0,0 +1,520 @@
+the task back on the executors queue of ready tasks, much like a thread
+scheduler in an operating system.
+
+Attempting to complete a task (or async value within it) is called *polling*,
+and always yields a `Poll` value back:
+
+```rust
+/// Indicates whether a value is available, or if the current task has been
+/// scheduled for later wake-up instead.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub enum Poll<T> {
+    /// Represents that a value is immediately ready.
+    Ready(T),
+
+    /// Represents that a value is not ready yet.
+    ///
+    /// When a function returns `Pending`, the function *must* also
+    /// ensure that the current task is scheduled to be awoken when
+    /// progress can be made.
+    Pending,
+}
+```
+
+When a task returns `Poll::Ready`, the executor knows the task has completed and
+can be dropped.
+
+### Waking up
+
+Each task executor provides its own scheduling facilities, and hence needs to
+customize the way task wakeups are handled. Most of the time, you should use the
+`std::task::Wake` trait defining wakeup behavior:
+
+```rust
+/// A way of waking up a specific task.
+///
+/// Any task executor must provide a way of signaling that a task it owns
+/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
+/// type that implements this trait.
+pub trait Wake: Send + Sync {
+    /// Indicates that the associated task is ready to make progress and should
+    /// be `poll`ed.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake` should place
+    /// the associated task onto this queue.
+    fn wake(self: &Arc<Self>);
+
+    /// Indicates that the associated task is ready to make progress and should be polled.
+    /// This function is like wake, but can only be called from the thread on which this
+    /// `Wake` was created.
+    ///
+    /// Executors generally maintain a queue of "ready" tasks; `wake_local` should place
+    /// the associated task onto this queue.
+    unsafe fn wake_local(self: &Arc<Self>)
+}
+```
+
+To see how this might be used in practice, here's a simple example sketch:
+
+```rust
+struct ExecutorInner {
+    sync_ready_queue: SynchronousQueue,
+    optimized_queue: UnsafeCell<Vec<Task>>,
+}
+
+struct Task {
+    future: ...,
+    executor: Arc<ExecutorInner>,
+}
+
+impl Wake for Task {
+    fn wake(self: &Arc<Self>) {
+        self.executor.sync_ready_queue.push(self.clone());
+    }
+    unsafe fn wake_local(self: &Arc<Self>) {
+        (&mut *self.executor.optimized_queue.get()).push(self.clone())
+    }
+}
+```
+
+The use of `&Arc<Self>` rather than just `&self` makes it possible to work directly with
+the trait object for `Wake`, including cloning it. With `UnsafeWake` below, we'll see
+an API with greater flexibility for the cases where `Arc` is problematic.
+
+In general async values are not coupled to any particular executor, so we use trait
+objects to handle waking. These come in two forms: `Waker` for the general case, and
```

Comment:
These don't look like trait objects to me.

Reply:
They're trait objects internally.